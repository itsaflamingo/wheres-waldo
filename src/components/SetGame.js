import '../styles/SetGame.css';
import React, {useEffect, useState} from 'react';
import GameOne from './GameOne';
import {saveCharacters, retrieve} from './firebaseConfig';
import MakeCursor from './MakeCursor';

function SetGame(props) {

    const {game} = props;

    const [background, setBackground] = useState({})
    const [displayCharacters, setDisplayCharacters] = useState(false);
    const [pointer, setPointer] = useState({});
    const [charactersFound, setCharactersFound] = useState([]);
    const [showGame, setShowGame] = useState(false);
    const [showInput, setShowInput] = useState(false);
    const [message, setMessage] = useState('');
    const [showLeaderboard, setShowLeaderboard] = useState(false);
    const [showMessage, setShowMessage] = useState(false);
    const [input, setInput] = useState({name: '', timer: 0});
    const [scores, setScores] = useState([{
      name: 'lisa',
      time: 0.04,
      id: 1 
    }, {
      name: 'cookie',
      time: 0.06,
      id: 2
    },{
      name: 'toki',
      time: 0.06,
      id: 3 
  }]);

    useEffect(() => {
      if(game.name === undefined) return;
      setBackground(game);
      setShowGame(true);
    })

    useEffect(() => {
      if(showGame === false) return;
      background.characters.map((char) => saveCharacters(char))
  
    }, [showGame])

    // Cursor movement
    useEffect(() => {
      const moveCursor = (e) => {
        const cursor = document.querySelector('#cursor');

        const x = e.pageX - 110;
        const y = e.pageY - 110;

        cursor.style.top = y + 'px';
        cursor.style.left = x + 'px';
      }
      document.addEventListener('mousemove', e => moveCursor(e))
      return () => {
        document.addEventListener('mousemove', e => moveCursor(e))
      }
    })

    // Display character option div when page is clicked
    useEffect(() => {
      if(background.image === undefined || showInput === true) return;
      const charDiv = document.querySelector('.character-options');
      // const backG = document.querySelector('#background');
      const setDivPosition = (e) => {

        e.stopPropagation();

        if(displayCharacters === false) return;
        const x = e.pageX;
        const y = e.pageY;
        
        charDiv.style.top = y + 'px';
        charDiv.style.left = x + 'px';
      }

      const getChosenCharacter = async (e) => {
        const id = e.target.id;
        const chosenCharacter = background.characters.filter(char => char.name === id);

        const dbChar = await retrieve(chosenCharacter[0].name);
        console.log(dbChar);
        checkIfCharacter(dbChar, id);
      }

      document.addEventListener('click', e => setDivPosition(e));
      if(charDiv === null) return() => {document.removeEventListener('click', e => setDivPosition(e))}
      charDiv.addEventListener('click', e => getChosenCharacter(e));

      return()=> {
        if (charactersFound.length < 3) return;
        document.removeEventListener('click', e => setDivPosition(e));
        charDiv.removeEventListener('click', e => getChosenCharacter(e));
      }
    }, [displayCharacters])

    useEffect(() => {
      if(charactersFound[charactersFound.length-1] === undefined) return;
      setMessage(() => `You found ${charactersFound[charactersFound.length-1]}!`)
      setShowMessage(true);
      setTimeout(() => {
        setShowMessage(false);
      }, 1500);
      if(charactersFound.length === 3) setShowInput(true);

    }, [charactersFound])

    
    const toggleDisplay = () => setDisplayCharacters(!displayCharacters);

    const handleMouseMove = (e) => {
      setPointer({
          x: e.pageX - e.target.offsetLeft,
          y: e.pageY - e.target.offsetTop
      })
  }

    const checkIfCharacter = (obj, name) => {
      const location = obj[name].location;

      console.log((pointer.x + 120) > location.x,
      (pointer.x - 120) < location.x, 
      (pointer.y + 120) > location.y,
      (pointer.y - 120) < location.y)

      if(((pointer.x + 120) > location.x &&
          (pointer.x - 120) < location.x) && 
          (pointer.y + 120) > location.y &&
          (pointer.y - 120) < location.y) {
        checkIfFound(name);
        console.log(`You found ${name}!`)
      }
    }

    const checkIfFound = (name) => {
      if(charactersFound.includes(name));
      setCharactersFound(() => charactersFound.concat(name));
    }

    useEffect(() => {
      if(scores.length === 3) return;
      setShowInput(false);
      setShowLeaderboard(true);
    }, [scores])

  const onChange = (e) => {
    const seconds = document.querySelector('#timer').firstChild.innerHTML;
    const milliseconds = document.querySelector('#timer').lastChild.innerHTML;
      setInput({name: e.target.value,
                time: `${seconds}${milliseconds}`,
                id: 4});
    }

  return (
    <div className="App">
      <MakeCursor />
      {showGame && (<GameOne 
                    background={background} 
                    displayCharacters={displayCharacters}
                    toggleDisplay={toggleDisplay}
                    handleMouseMove={handleMouseMove}
                    message={message} 
                    showInput={showInput} 
                    showLeaderboard={showLeaderboard} 
                    setShowLeaderboard={setShowLeaderboard} 
                    scores={scores}
                    setScores={setScores} 
                    showMessage={showMessage}
                    onChange={onChange} 
                    input={input} /> 
                    )}
    </div>
  );
}

export default SetGame;
