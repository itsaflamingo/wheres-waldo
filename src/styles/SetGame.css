body {
    margin: 0;
    cursor: none;
    --gradient: linear-gradient(90deg, rgb(197, 25, 25), rgb(195, 107, 14));
    --box-shadow: 0 0px 50px rgba(255, 255, 255, 0.558),
                  inset 0 0 20px rgba(250, 245, 245, 0.558);
    --yellow: rgb(252, 181, 29);
    font-family: sans-serif;
}
.App {
    width: max-content;
    height: max-content;
}
#background {
    position: relative;
}
#character-options {
    position: absolute;
    background-color: white;
    height: 1%;
    width: 12%;
    font-size: 1.5rem;
    border-radius: 20px;
}
#character-options:hover {
    cursor: pointer;
}
.character-options {
    position: absolute;
    background-color: white;
    font-size: 1.5rem;
    border-radius: 20px;
}
.character-options .character {
    height: 33%;
    display: flex;
    align-items: center;
    padding-left: 10%;
    border-bottom: 1px solid grey;
}
[id='Robot City'] {
    height: 2.5%;
    width: 10%;
}
[id='Universe 113'] {
    height: 4%;
    width: 10%;
}
[id='A.D.2222'] {
    z-index: 10;
    height: 1%;
    width: 12%;
}
.character:hover {
    background-color: grey;
}
/* HEADER */
#header {
    position: fixed;
    top: 0;
    width: 90vw;
    z-index: 10;
    display: flex;
    justify-content: space-between;
    background-color: transparent;
    padding-top: 1.5%;
    padding-left: 5%;
    padding-right: 5%;
}
#message {
    position: absolute;
    left: 45%;
    color: white;
}
#timer {
    font-size: 40px;
    color: white;
    font-weight: 900;
}
#show-leaderboard {
    border-radius: 20px;
    border: none;
    width: 10%;
    min-width: 107px;
    font-size: 1rem;
    font-weight: 900;
    background-color: var(--yellow);
}
#leaderboard {
    background-color: white;
    width: 20%;
    border-radius: 20px;
    height: auto;
    position: absolute;
    top: 100%;
    left: 40%;
}
.score {
    display: flex;
    justify-content: space-between;
    margin: 10% 10%;
    border-bottom: 1px solid grey;
    height: 20%;
}
.score > * {
    display: flex;
    align-items: center;
    justify-content: center;
}
#title {
    display: flex;
    justify-content: space-between;
    padding-bottom: 7%;
    margin: 0 10%;
    border-bottom: 1px solid grey;
}
#title h4{
    margin: 10% 0;
}
/* INPUT */
#enter-name {
    position: fixed;
    top: 45%;
    left: 45%;
    background-color: white;
    height: 15%;
    width: 25%;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    border-radius: 20px;
    z-index: 100;
    cursor: auto;
}
button[type=submit] {
    margin-top: 5%;
    border: none;
    height: 35%;
    width: 35%;
    background-color: var(--yellow);
    border-radius: 20px;
    font-weight: 900;
}
/* CURSOR */
#cursor {
    height: 220px;
    width: 220px;
    border-radius: 50%;
    border: 4px dashed white;
    position: absolute;
    pointer-events: none;
    z-index: 1;
}
.no-events {
    pointer-events: none;
}
.no-events > #character-options {
    pointer-events: auto;
}
/* CHOOSE GAME */
#page {
    background-image: var(--gradient);
    height: 100vh;
    width: 100vw;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    cursor: auto;
    color: rgb(44, 22, 0);
}
#start-btn {
    width: 100%;
    height: 10%;
    display: flex;
    align-items: center;
    justify-content: center;
}
#start {
    height: 70%;
    width: 10%;
    border-radius: 10px;
    border: none;
    background-color: var(--yellow);
    font-size: 1.5rem;
    font-weight: 900;
    margin-bottom: 3%;
}
#start:hover, #show-leaderboard:hover, button[type=submit]:hover {
    box-shadow: var(--box-shadow);
    background-color: rgb(3, 126, 208);
    color: rgb(251, 228, 183);
}
#choose-game {
    height: 100%;
    width: 100%;
    display: flex;
    align-items: center;
    justify-content: center;
    gap: 0.5%;
}
.game {
    display: flex;
    width: 90%;
    height: 85%;
    background-color: white;
    align-items: center;
    justify-content: center;
    border-radius: 10%;
    background-image: var(--gradient);
}
.game:hover {
    box-shadow: var(--box-shadow);
}
.game-container {
    height: 95%;
    width: 95%;
    display: flex;
    flex-direction: column;
    align-items: center;
}
.game > * {
    pointer-events: none;
}
.content {
    height: 90%;
    width: 90%;
    display: flex;
    align-items: center;
    justify-content: center;
    border-radius: 10%;
    background-color: rgba(252, 221, 221, 0.303);
}
.title {
    margin: 2% 0;
}
.background {
    height: 80%;
    width: 50%;
    border-radius: 20px;
}
.characters {
    height: 80%;
    width: 33%;
}
.character-div {
    display: flex;
    height: 40%;
    width: 100%;
}
.character {
    width: 90%;
    height: 90%;
}
.name {
    display: flex;
    align-items: center;
    justify-content: center;
    height: 55%;

}
