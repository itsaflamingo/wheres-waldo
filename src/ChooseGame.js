import React, { useState } from "react";
import './styles/SetGame.css';
import AliceCarousel from 'react-alice-carousel';
import "react-alice-carousel/lib/alice-carousel.css";
// import SetGame from "./components/SetGame.js";
import MakeBackgrounds from "./components/Backgrounds";
import ChooseGameDiv from "./components/ChooseGameDiv";

function ChooseGame() {

    const [showGame, setShowGame] = useState(false);
    const [game, SetGame] = useState({});
    const [games, setGames] = useState(MakeBackgrounds());

    const startGame = (e) => {
        setShowGame(true)
    }

    return(
        <div id='page'>
                <ChooseGameDiv games={games} />
                <div id='startBtn'>
                    <button id='start'
                    onClick={(e) => startGame(e)}>Start</button>
            </div>
            {/* {showGame && <SetGame game={game} />} */}
        </div>
    )
}

export default ChooseGame;