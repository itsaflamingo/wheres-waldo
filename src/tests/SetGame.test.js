import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";  // optional
import userEvent from "@testing-library/user-event";
import {BrowserRouter as Router} from 'react-router-dom';
import ChooseGame from '../ChooseGame'
import SetGame from '../components/SetGame';
import MakeBackgrounds from '../components/Backgrounds'
import { act } from "react-dom/test-utils";

describe('Game page', () => {
    const games = MakeBackgrounds();
    const game = games[0]
    it('Renders game page', () => {
        const {container} = render(<SetGame game={game} />)
        expect(container).toMatchSnapshot();
    })
    it('Background must have src attribute with correct image', () => {
        render(<SetGame game={game} />)
    
        expect(screen.getByRole('img')).toHaveAttribute('src', `${game.image}`)
    })
    it('When clicked, displays character list', () => {
        render(<SetGame game={game} />)

        const background = screen.getByTestId('background')
        
        userEvent.click(background);

        const HK = screen.queryByText(/Hollow Knight/i);
        const Yubaba = screen.queryByText(/Yubaba/i);
        const Bowser = screen.queryByText(/Bowser/i);

        expect(HK).not.toBeNull()
        expect(Yubaba).not.toBeNull()
        expect(Bowser).not.toBeNull()

    })
    // When character is found, display message after setTimeout
    // Timer updates every second
})
