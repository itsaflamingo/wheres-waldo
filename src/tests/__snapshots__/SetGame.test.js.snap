// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game page Renders game page 1`] = `
<div>
  <div
    class="App"
  >
    <div
      id="cursor"
    />
    <div
      data-testid="background"
      id="background"
    >
      <div
        id="header"
      >
        <div
          id="timer"
        >
          <span>
            00
            :
          </span>
          <span>
            00
          </span>
        </div>
        <button
          id="show-leaderboard"
        >
          Leaderboard
        </button>
      </div>
      <img
        alt="puzzle"
        src="img1.jpg"
      />
    </div>
  </div>
</div>
`;
